### PR TITLE
add price and meter dynamically add foreign keys in billing

### PR DESCRIPTION
**Pull request by anamarn**
**TLDR**
Solves: https://github.com/twentyhq/private-issues/issues/199
Partially solves: https://github.com/twentyhq/private-issues/issues/221 (more details below)

Updates the BillingMeter and BillingPrice tables while listening to the events "price.created" and "price.updated" from the stripe webhook. Also added the foreign keys, that couldn't be added to the BillingEntities.

**In Order To test**
Billing:

- Set IS_BILLING_ENABLED to true
- Add your BILLING_STRIPE_SECRET and BILLING_STRIPE_API_KEY
- Add your BILLING_STRIPE_BASE_PLAN_PRODUCT_ID (use the one in testMode > Base Plan)

Authenticate with your account in the stripe CLI
Run the command: stripe listen --forward-to http://localhost:3000/billing/webhooks

Run the twenty workker

Authenticate yourself on the app choose a plan and run the app normally. In stripe and in posgress the customer table data should be added.

**Take Into Consideration**

In a previous migration the foreign key to workpaceId was taken down this was due to the separation of the migrations if billing is enabled. Because we want to separate in these two categories: we will be polluting the Common Migrations with relations to tables that don't exists. This will be addressed in a PR in the next sprint (perhaps a decorator?)

**Doing**
Testing migrations, when we are in main and when billing is enabled.

### CODE REVIEW

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      `ALTER TABLE "core"."billingSubscriptionItem" DROP COLUMN "quantity"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" ADD "quantity" numeric`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingCustomer" DROP CONSTRAINT "IndexOnWorkspaceIdAndStripeCustomerIdUnique"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingCustomer" ADD CONSTRAINT "UQ_53c2ef50e9611082f83d760897d" UNIQUE ("workspaceId")`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts:25]: give better name

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  implements MigrationInterface
+{
+  name = 'AddForeignKeysIndexOnBilling1734369352850';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" DROP CONSTRAINT "IndexOnBillingSubscriptionIdAndStripeSubscriptionItemIdUnique"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" ADD CONSTRAINT "UQ_6a989264cab5ee2d4b424e78526" UNIQUE ("stripeSubscriptionItemId")`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts:13]: better name

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      `ALTER TABLE "core"."billingCustomer" DROP CONSTRAINT "IndexOnWorkspaceIdAndStripeCustomerIdUnique"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingCustomer" ADD CONSTRAINT "UQ_53c2ef50e9611082f83d760897d" UNIQUE ("workspaceId")`,
+    );
+    await queryRunner.query(
+      `CREATE UNIQUE INDEX "IndexOnActiveSubscriptionPerWorkspace" ON "core"."billingSubscription" ("workspaceId") WHERE status IN ('trialing', 'active', 'past_due')`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" ADD CONSTRAINT "IndexOnBillingSubscriptionIdAndStripeSubscriptionItemIdUnique" UNIQUE ("billingSubscriptionId", "stripeSubscriptionItemId")`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts:31]: this constraint is not useful, maybe we want one on billingSubscriptionId + update upsert

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingCustomer" ADD CONSTRAINT "IndexOnWorkspaceIdAndStripeCustomerIdUnique" UNIQUE ("workspaceId", "stripeCustomerId")`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts:34]: same here

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingEntitlement" ADD CONSTRAINT "FK_766a1918aa3dbe0d67d3df62356" FOREIGN KEY ("stripeCustomerId") REFERENCES "core"."billingCustomer"("stripeCustomerId") ON DELETE CASCADE ON UPDATE NO ACTION`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscription" ADD CONSTRAINT "FK_9120b7586c3471463480b58d20a" FOREIGN KEY ("stripeCustomerId") REFERENCES "core"."billingCustomer"("stripeCustomerId") ON DELETE NO ACTION ON UPDATE NO ACTION`,
```
[the remark this answers is not in the document]
Comment (charlesBochet) [reply packages/twenty-server/src/database/typeorm/core/migrations/billing/1734369352850-addForeignKeysIndexOnBilling.ts:40]: CASCADE DELETE

File: packages/twenty-server/src/engine/core-modules/billing/stripe/stripe.service.ts [hunk 6]
```diff
@@ -146,7 +143,7 @@ export class StripeService {
       stripeSubscriptionItem.stripeSubscriptionItemId,
       {
         price: stripePriceId,
-        quantity: stripeSubscriptionItem.quantity,
+        quantity: stripeSubscriptionItem.quantity ?? undefined,
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/stripe/stripe.service.ts:146]: is this useful?

File: packages/twenty-server/src/engine/core-modules/billing/utils/transform-stripe-price-event-to-price-repository-data.util.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
+import Stripe from 'stripe';
+
+import { BillingPriceBillingScheme } from 'src/engine/core-modules/billing/enums/billing-price-billing-scheme.enum';
+import { BillingPriceTaxBehavior } from 'src/engine/core-modules/billing/enums/billing-price-tax-behavior.enum';
+import { BillingPriceTiersMode } from 'src/engine/core-modules/billing/enums/billing-price-tiers-mode.enum';
+import { BillingPriceType } from 'src/engine/core-modules/billing/enums/billing-price-type.enum';
+import { SubscriptionInterval } from 'src/engine/core-modules/billing/enums/billing-subscription-interval.enum';
+import { BillingUsageType } from 'src/engine/core-modules/billing/enums/billing-usage-type.enum';
+
+export const transformStripePriceEventToPriceRepositoryData = (
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/utils/transform-stripe-price-event-to-price-repository-data.util.ts:10]: add unit tests

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      `ALTER TABLE "core"."billingSubscriptionItem" ADD CONSTRAINT "UQ_6a989264cab5ee2d4b424e78526" UNIQUE ("stripeSubscriptionItemId")`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" DROP COLUMN "quantity"`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingSubscriptionItem" ADD "quantity" numeric`,
+    );
+    await queryRunner.query(
+      `ALTER TABLE "core"."billingCustomer" ADD CONSTRAINT "UQ_53c2ef50e9611082f83d760897d" UNIQUE ("workspaceId")`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts:25]: I think we also want a constraint on stripeCustomerId

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    );
+    await queryRunner.query(
+      `CREATE UNIQUE INDEX "IndexOnActiveSubscriptionPerWorkspace" ON "core"."billingSubscription" ("workspaceId") WHERE status IN ('trialing', 'active', 'past_due')`,
```
Comment (charlesBochet) [inline packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts:28]: I think we also want a unique constraint on billingSubscriptionStripeId

**Review by charlesBochet — APPROVED** [review]
Thank you, i've left comments that can be treated in an upcoming PR

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts (hunk 8 above)
Comment (anamarn) [reply packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts:25]: it's already implemented on the creation of the table Billing Customer

File: packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts (hunk 9 above)
Comment (anamarn) [reply packages/twenty-server/src/database/typeorm/core/migrations/billing/1734450749954-addConstraintsOnBillingTables.ts:28]: This constraint has been added in the first billing migration. 

Am I missing something?
